### PR TITLE
add state and block overrides in `eth_call`

### PR DESCRIPTION
**Pull request by tcoratger**
Time spent on this PR:

Resolves: #<Issue number>

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

# Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/tracing/database.rs [hunk 1]
```diff
@@ -11,9 +11,9 @@ use tokio::runtime::Handle;
 
 #[derive(Debug)]
 #[allow(clippy::redundant_pub_crate)]
-pub(crate) struct EthDatabaseSnapshot<P: EthereumProvider + Send + Sync> {
-    cache: CacheDB<P>,
-    block_id: BlockId,
+pub struct EthDatabaseSnapshot<P: EthereumProvider + Send + Sync> {
```
Comment (tcoratger) [inline src/tracing/database.rs:14]: Here I didn't want to move all that in this PR as it is already quite big but I think that in a follow up PR it will be necessary to remove this snapshot database file from the tracing to put an equivalent file in the provider given that we is used for the state overrides of this snapshot.

File: tests/tests/eth_provider.rs [hunk 2]
```diff
@@ -809,6 +812,179 @@ async fn test_send_raw_transaction_wrong_signature(#[future] katana: Katana, _se
...
+            EthApiError::Transaction(TransactionError::Overrides(err)) => {
+                assert_eq!(
+                    err.to_string(),
+                    "transaction validation error: lack of funds (1000000000) for max fee (1000210001)",
+                );
+            }
+            // If a different error occurs, panic
+            _ => panic!("Expected a transaction error but got a different error: {e:?}"),
+        },
+    }
```
Comment (greged93) [inline tests/tests/eth_provider.rs:900]: This can be removed I think

File: tests/tests/eth_provider.rs [hunk 3]
```diff
@@ -809,6 +812,179 @@ async fn test_send_raw_transaction_wrong_signature(#[future] katana: Katana, _se
...
+    };
+
+    // Perform the call to invoke the incForLoop function without state override
+    let _ = eth_provider
+        .call(request3, None, None, None)
+        .await
+        .expect("Failed to call incForLoopCall in Kakarot constract");
+
+    // Define another Solidity contract with a different interface and bytecode
+    sol! {
```
Comment (greged93) [inline tests/tests/eth_provider.rs:932]: this test is very nice, I would basically:
- generate a random address and fund it via state override.
- generate the random bytecode as you did.
- call the contract using the random generated address.

TLDR: only keep this lower test as the override test and move the previous test to an `eth_call` test

File: src/eth_provider/error.rs [hunk 4]
```diff
@@ -285,6 +305,9 @@ pub enum TransactionError {
     /// Thrown if the tracing fails
     #[error("tracing error: {0}")]
     Tracing(Box<dyn std::error::Error + Send + Sync>),
+    /// Thrown if the call with state or block overrides fails
+    #[error("tracing error: {0}")]
+    Overrides(Box<dyn std::error::Error + Send + Sync>),
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/eth_provider/error.rs]: Agreed, just fixed

File: src/eth_provider/provider.rs [hunk 5]
```diff
@@ -160,6 +181,41 @@ where
...
 
+impl<SP> DatabaseRef for EthDataProvider<SP>
+where
+    SP: starknet::providers::Provider + Send + Sync,
+{
+    type Error = EthApiError;
+
+    fn basic_ref(&self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        tokio::task::block_in_place(|| {
+            let mut database_snapshot = EthDatabaseSnapshot::new(self, BlockId::default());
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/eth_provider/provider.rs:192]: block id by default is latest, this is what we should take here no?

File: src/eth_provider/provider.rs [hunk 6]
```diff
@@ -160,6 +181,41 @@ where
     }
 }
 
+impl<SP> DatabaseRef for EthDataProvider<SP>
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/eth_provider/provider.rs:184]: Yes this is needed for this constraint https://github.com/paradigmxyz/reth/blob/7bf99d8a736b62d05aa00d1e7e7d7b9fa11e0346/crates/rpc/rpc-eth-api/src/helpers/call.rs#L884

in our `prepare_call_env` call and not on `EthDatabaseSnapshot`...

File: tests/tests/eth_provider.rs (hunk 3 above)
Comment (tcoratger) [reply tests/tests/eth_provider.rs:964]: Yes I agree with you. I wanted to do a single test to avoid restarting Katana several times but ultimately it is too big and these tests are too important to neglect.

So I separated into several tests:
- One without any state override just to check that a simple transfer passes
- One with state override on transfer with success because the random account has a balance **greater** than what must be transferred.
- One with state override on transfer with failure because the random account has a balance **lower** than what must be transferred.
- One with modification of the bytecode after a first successful call to be sure that the state override overwrites the settings
- One with modification of the bytecode and call to the new bytecode to be sure that everything works well on a call after state overrides.


Normally, block overrides are also automatically included but not a priority so we do not test them on this PR which is already quite big. I suggest opening another issue solely dedicated to testing block overrides.

File: tests/tests/eth_provider.rs [hunk 7]
```diff
@@ -809,6 +812,179 @@ async fn test_send_raw_transaction_wrong_signature(#[future] katana: Katana, _se
...
+    let katana = plain_opcodes.0;
+
+    // Obtain an Ethereum provider instance from the Katana instance
+    let eth_provider = katana.eth_provider();
+
+    // Convert KakarotEvmContract's EVM address to an Address type
+    let contract_address = Address::from_slice(&plain_opcodes.1.evm_address.to_bytes_be()[12..]);
+
+    // Get the EOA (Externally Owned Account) address from Katana
+    let eoa_address = katana.eoa().evm_address().expect("Failed to get eoa address");
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply tests/tests/eth_provider.rs]: Yes just changed the logic here

File: tests/tests/eth_provider.rs [hunk 8]
```diff
@@ -809,6 +812,179 @@ async fn test_send_raw_transaction_wrong_signature(#[future] katana: Katana, _se
...
+    };
+
+    // Call the Ethereum provider with the created transaction request
+    let _ = eth_provider.call(request, None, None, None).await.expect("Failed to call");
+}
+
+#[rstest]
+#[awt]
+#[tokio::test(flavor = "multi_thread")]
+async fn test_call_with_state_override(#[future] plain_opcodes: (Katana, KakarotEvmContract), _setup: ()) {
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply tests/tests/eth_provider.rs]: agreed and done

File: Makefile [hunk 9]
```diff
@@ -58,6 +58,9 @@ test: katana-genesis load-env
 test-target: load-env
 	cargo test --tests --features testing $(TARGET) -- --nocapture
 
+test-target1: load-env
+	cargo test --package kakarot-rpc --test entry --all-features -- tests::eth_provider::test_call_with_state_override_bytecode1 --exact --show-output
+
 benchmark:
```
Comment (tcoratger) [inline Makefile:64]:
```suggestion

benchmark:
```

File: src/eth_provider/database/state.rs [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+                    .unwrap_or_default();
+                Result::<_, EthApiError>::Ok(hash)
+            })?;
+
+            Ok(hash)
+        })
+    }
+}
+
+impl<P: EthereumProvider + Send + Sync> Database for EthCacheDatabase<P> {
```
Comment (greged93) [inline src/eth_provider/database/state.rs:98]: I don't think we need all this, pretty sure `CacheDB<T>` already implements `Database` and `DatabaseRef` if `T` implements `Database` and `DatabaseRef`

File: src/eth_provider/database/state.rs [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+        let hash = DatabaseRef::block_hash_ref(&self.0, block_number)?;
+        self.0.block_hashes.insert(number, hash);
+
+        Ok(hash)
+    }
+}
+
+impl<P: EthereumProvider + Send + Sync> DatabaseCommit for EthCacheDatabase<P> {
+    fn commit(&mut self, changes: HashMap<Address, Account>) {
```
Comment (greged93) [inline src/eth_provider/database/state.rs:158]: same, I think `CachedDB` implements it already

File: tests/tests/eth_provider.rs [hunk 12]
```diff
@@ -805,6 +808,260 @@ async fn test_send_raw_transaction_wrong_signature(#[future] katana: Katana, _se
...
+            // If a different error occurs, panic
+            _ => panic!("Expected an invalid transaction revert error but got a different error: {e:?}"),
+        },
+    }
+}
+
+#[rstest]
+#[awt]
+#[tokio::test(flavor = "multi_thread")]
+async fn test_call_with_state_override_bytecode1(#[future] plain_opcodes: (Katana, KakarotEvmContract), _setup: ()) {
```
Comment (greged93) [inline tests/tests/eth_provider.rs:1006]: I'm not sure we need this extract test, I don't think it brings much more over the previous test

Comment (tcoratger) [reply tests/tests/eth_provider.rs:1006]: Mhhh not sure, maybe better to delete the first one but in my opinion, knowing that it's quite complex and that it doesn't add too much cost I would keep both.
- The firs tests that if I deploy a smart contract and I call a function inside it works (normal). But if I override the smart contract with another bytecode then this same call no longer works (the override has taken place).
- This second test that you point out is a little different because it verifies that if I deploy a smart contract and I change the bytecode directly to mimic another smart contract then this time I have the right to call a function of this new smart contract and it works (the override worked well and the bytecode read corresponds to what I want for my new smart contract).

So I would say that the second test is more powerful and if I had to keep only one I would keep this one. As you prefer, if you prefer to remove the first one for CI costs then let's do it :)

Comment (greged93) [reply tests/tests/eth_provider.rs:1006]: I don't really see the difference in the two tests, for me it's just testing that override works, once with a call before to the same contract and the second test without this extra test. I would remove at least one

Comment (tcoratger) [reply tests/tests/eth_provider.rs]: Just removed the first one, as the second more powerful (testing the call to a new function of the override).

File: src/eth_provider/database/state.rs (hunk 10 above)
Comment (tcoratger) [reply src/eth_provider/database/state.rs]: Oh yes really nice you are right, just fixed

File: src/eth_provider/database/state.rs (hunk 11 above)
Comment (tcoratger) [reply src/eth_provider/database/state.rs]: Oh yes really nice you are right, just fixed

File: Makefile [hunk 13]
```diff
@@ -58,6 +58,9 @@ test: katana-genesis load-env
 test-target: load-env
 	cargo test --tests --features testing $(TARGET) -- --nocapture
 
+test-target1: load-env
+	cargo test --package kakarot-rpc --test entry --all-features -- tests::trace_api::test_trace_block --exact --show-output
+
 benchmark:
```
Comment (tcoratger) [inline Makefile:64]:
```suggestion
test-target: load-env
	cargo test --tests --features testing $(TARGET) -- --nocapture

benchmark:
```

File: src/eth_provider/provider.rs [hunk 14]
```diff
@@ -433,7 +454,66 @@ where
...
+                prevrandao: Some(B256::from_slice(&difficulty.to_be_bytes::<32>()[..])),
+                ..Default::default()
+            };
+
+            // Combine the configuration environment with the handler configuration.
+            let cfg_env_with_handler_cfg =
+                CfgEnvWithHandlerCfg { cfg_env, handler_cfg: HandlerCfg::new(SpecId::CANCUN) };
+
+            // Create a snapshot of the Ethereum database using the block ID.
+            let mut db = EthCacheDatabase(CacheDB::new(EthDatabase::new(self, block_id.unwrap_or_default())));
```
Comment (greged93) [inline src/eth_provider/provider.rs:492]: I don't know here if we need to pass the block_id or the parent_block id, have you checked in Reth?

File: tests/tests/eth_provider.rs [hunk 15]
```diff
@@ -805,6 +807,170 @@ async fn test_send_raw_transaction_wrong_signature(#[future] katana: Katana, _se
...
+            EthApiError::Transaction(TransactionError::Call(err)) => {
+                assert_eq!(
+                    err.to_string(),
+                    "transaction validation error: lack of funds (1000000000) for max fee (1000210001)",
+                );
+            }
+            // If a different error occurs, panic
+            _ => panic!("Expected a transaction error but got a different error: {e:?}"),
+        },
+    }
```
Comment (greged93) [inline tests/tests/eth_provider.rs:909]: can be replaced by something shorter no?
```rust
let res = eth_provider.call(request, None, Some(state_override.clone()), None).await;
let err = res.unwrap_err().to_string();
assert_eq!(err, "transaction validation error: lack of funds (1000000000) for max fee (1000210001)");
```

Comment (tcoratger) [reply tests/tests/eth_provider.rs:909]: If you agree as it is a test I would like to leave these different messages like that, even if the test fails, I have an indication of what is happening in the CI (for example another error than the expected one or the fact that I expected a failure but I got a success). I think it can help us understand an error more quickly if one day we have a problem in the test. wdyt?

Comment (greged93) [reply tests/tests/eth_provider.rs:909]: > (for example another error than the expected one or the fact that I expected a failure but I got a success)

I agree but wouldn't `unwrap_err` and the `assert_eq` give you both of this?

File: src/tracing/mod.rs [hunk 16]
```diff
@@ -1,9 +1,8 @@
 pub mod builder;
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/tracing/mod.rs:1]: Yes right, just did it

File: src/eth_provider/provider.rs (hunk 14 above)
Comment (tcoratger) [reply src/eth_provider/provider.rs:492]: I think this is the block ID as far as I understand with:

https://github.com/paradigmxyz/reth/blob/ed16643f122fcb010ae6744bee72e469d6692af1/crates/rpc/rpc-eth-api/src/helpers/call.rs#L360-L362

File: tests/tests/eth_provider.rs (hunk 15 above)
Comment (tcoratger) [reply tests/tests/eth_provider.rs:909]: Yes in theory should be, convinced, let us simplify and if we have issues wit debugging we can revert in the future :)

File: src/eth_provider/error.rs [hunk 17]
```diff
@@ -98,16 +105,29 @@ impl std::fmt::Display for EthApiError {
...
-            EthApiError::Execution(ExecutionError::Evm(EvmError::Other(ref b))) => Some(b),
+            EthApiError::Execution(ExecutionError::Evm(EvmError::Other(ref b))) => Some(b.clone()),
             _ => None,
         };
         ErrorObject::owned(code as i32, msg, data)
     }
 }
 
+impl ToRpcError for EthApiError {
+    fn to_rpc_error(&self) -> ErrorObject<'static> {
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/eth_provider/error.rs:126]: Yes it is required by:

```rust
impl From<EthApiError> for RethEthApiError {
    fn from(value: EthApiError) -> Self {
        Self::other(value)
    }
}
```

Otherwise:

```
error[E0277]: the trait bound `eth_provider::error::EthApiError: ToRpcError` is not satisfied
   --> src/eth_provider/error.rs:50:21
    |
50  |         Self::other(value)
    |         ----------- ^^^^^ the trait `ToRpcError` is not implemented for `eth_provider::error::EthApiError`
    |         |
    |         required by a bound introduced by this call
    |
note: required by a bound in `reth_rpc_eth_types::EthApiError::other`
   --> /Users/tcoratger/.cargo/git/checkouts/reth-36d3ea1d1152b20c/d599393/crates/rpc/rpc-eth-types/src/error.rs:137:21
    |
137 |     pub fn other<E: ToRpcError>(err: E) -> Self {
    |                     ^^^^^^^^^^ required by this bound in `EthApiError::other`
```

File: src/eth_provider/provider.rs [hunk 18]
```diff
@@ -450,7 +471,66 @@ where
...
+                cfg_env_with_handler_cfg,
+                block_env,
+                request.clone(),
+                request.gas.unwrap_or_default().try_into().expect("Gas limit is too large"),
+                &mut db.0,
+                evm_overrides,
+            )?;
+
+            // Execute the transaction using the configured EVM.
+            let res = tokio::task::block_in_place(|| {
```
Comment (greged93) [inline src/eth_provider/provider.rs:522]: why block in place here? is this needed?

File: src/eth_provider/database/state.rs [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+                Result::<_, EthApiError>::Ok(AccountInfo { nonce, balance, code: Some(bytecode), code_hash })
+            })?;
+
+            Ok(Some(account_info))
+        })
+    }
+
+    /// Returns the code for the given code hash.
+    /// TODO: Implement this method in the provider
```
Comment (greged93) [inline src/eth_provider/database/state.rs:64]: can be removed no?

File: src/eth_provider/provider.rs (hunk 18 above)
Comment (tcoratger) [reply src/eth_provider/provider.rs]: just fixed you are right

File: src/eth_provider/database/state.rs (hunk 19 above)
Comment (tcoratger) [reply src/eth_provider/database/state.rs:64]: As discussed, let us keep as is for now

**Review by greged93 — APPROVED** [review]
